File: src/ParticleEngine/ParticleEngine/Scripts/Particle/Physics/Materials/SimpleGravity.cs
using static ParticleEngine.Particle.ParticleManager;

namespace ParticleEngine.Particle.Physics.Materials
{
    public static class SimpleGravity
    {
        public static void Calculate(int row, int column, int rowSize, int columnSize, ref ParticleManager particleWorld)
        {
            ParticleInstance self = particleWorld.particles[row, column];

            if (row < rowSize - 1)
            {
                if (
                    (particleWorld.particles[row + 1, column].materialType == MaterialType.Gas ||
                    particleWorld.particles[row + 1, column].material == Material.Water) &&
                    particleWorld.particles[row, column].material == self.material)
                {
                    particleWorld.particles[row + 1, column] = self;
                    particleWorld.Clear(row, column);
                }
            }
        }
    }
}

File: src/ParticleEnginePlusPlus/ParticleEngine/ParticleEngine/include/particle/physics/materials/simple_gravity.cpp
#include <iostream>
#include "particle/particle_world.h"
#include "particle/particle_physics.h"

void calculate_simpleGravity(int row, int col, ParticleWorld* particleWorld)
{
	ParticleWorld::ParticleInstance self = particleWorld->getParticle(row, col);

	if (particleWorld->canDown(row))
	{
		if (
			(particleWorld->getParticle(row + 1, col).materialType == ParticleWorld::MaterialType::Gas ||
				particleWorld->getParticle(row + 1, col).material == ParticleWorld::Material::Water) &&
			particleWorld->getParticle(row, col).material == self.material)
		{
			particleWorld->setParticle(row + 1, col, self);
			particleWorld->resetParticle(row, col);
		}
	}
}